replicatedhq/ship
# Reworking some structure around lifecycle/state in the the daemon

What I Did
------------

Preliminary work for reworking navigation in ship. Not fully working yet. Basic idea is to
allow a user to navigate via the browser's forward/back buttons. 

Apologies in advance for the massive PR, and this probably should have been a REP first.

High level goals:

- give daemon more knowledge of lifecycle
- don't give daemon too much knowledge of lifecycle

How I Did it
------------

(in order of appearance)

Added an `id` field to every lifecycle step, and some interfaces to make
reading ids out a little easier. These ID's will be vendor-customizable, and map to url routes on the front end.
For a step with id `intro`, going to the route `/intro` will always render that step, regardless of the Daemon's `currentStep` field.


`PushXXXXStep` methods on daemon augmented to get the `api.Step`, so they know about the step ID.


We add two new endpoints `/lifecycle`, for client to 
get a list of all step ids and some info about what type of 
component should be used to render, e.g. `message`, `render`, `helm-intro`.

The other endpoint `/lifecycle/step/:stepID`, will function like `getCurrentStep`,
but will allow client to retrieve step info for any step *that has already been reached*. 
So no skipping steps, but you can navigate freely between steps you've already done.

Refactor out some logic from `d.getCurrentStep` so we can also use it in the method to get a specific step.

Confirm message actions now send the step ID, and the endpoint will quietly ignore a confirmation for a non-current message step.

remove `getCurrentMessage` endpoint, its unused

Add a `specs.IDPatcher` struct to add `ID`s to steps in a lifecycle if they don't have them via `EnsureAllStepsHaveUniqueIDs`.


Things we still have to figure out
------------

There are still some big holes in this, but in general it feels like the right direction.

- if you go "back" to the config screen, then save, do we run-run the render step?
  - other similar questions for render-y things
- FE integration still under way
- I'd really like to track some stuff about how/when steps have been completed in state.json
  - The interplay between lifecycle.Runner and Daemon is already a little complex, throwing more state responsibility into the mix 
    will likely make that even worse
- probably some other big holes



How to verify it
------------

Run ship, confirm everything is broken :sad:

(you also can try out the endpoints if you want)


Description for the Changelog
------------

None yet



![](https://www.featurepics.com/FI/Thumb300/20061103/Sinking-Ship-128527.jpg)
